langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java

```diff
@@ -0,0 +1,506 @@
+        return addAll(embeddings, null);
+    }
+
+    @Override
+    public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedded) {
+        List<DocumentChunk.Id> keys = generateKeys(embeddings.size());
+        addAllInternal(keys, embeddings, embedded);
+        return keys.stream().map(DocumentChunk.Id::toString).collect(Collectors.toList());
+    }
+
+    @Override
+    public void remove(String id) {
+        if (id == null || id.isBlank()) {
+            throw new IllegalArgumentException("id cannot be null or blank");
+        }
+
+        documentChunks.remove(DocumentChunk.Id.parse(id));
+    }
+
+    @Override
+    public void removeAll(Collection<String> ids) {
+        if (ids == null || ids.isEmpty()) {
+            throw new IllegalArgumentException("ids cannot be null or empty");
+        }
+
+        documentChunks.keySet().removeAll(ids.stream().map(DocumentChunk.Id::parse).collect(Collectors.toSet()));
+    }
+
+    @Override
+    public void removeAll(dev.langchain4j.store.embedding.filter.Filter filter) {
+        if (filter == null) {
+            throw new IllegalArgumentException("filter cannot be null");
+        }
+        
+        Filter<DocumentChunk> chunkFilter = CoherenceMetadataFilterMapper.map(filter);
+        documentChunks.invokeAll(chunkFilter, CacheProcessors.removeBlind());
+    }
+
+    @Override
+    public void removeAll() {
+        documentChunks.truncate();
+    }
+
+    @Override
+    public EmbeddingSearchResult<TextSegment> search(EmbeddingSearchRequest request) {
+        Embedding queryEmbedding = request.queryEmbedding();
+        if (normalizeEmbeddings) {
+            queryEmbedding.normalize();
+        }
+        double minScore = request.minScore();
+        boolean checkMin = minScore != 0.0f;
+        Filter<DocumentChunk> filter = CoherenceMetadataFilterMapper.map(request.filter());
+        Float32Vector vector = new Float32Vector(queryEmbedding.vector());
+
+        SimilaritySearch<DocumentChunk.Id, DocumentChunk, float[]> aggregator = new SimilaritySearch<>(EXTRACTOR, vector, request.maxResults());
+        if (filter != null) {
+            aggregator = aggregator.filter(filter);
+        }
+            
+        List<QueryResult<DocumentChunk.Id, DocumentChunk>> results = documentChunks.aggregate(aggregator);
+        List<EmbeddingMatch<TextSegment>> matches = new ArrayList<>();
+
+        for (QueryResult<DocumentChunk.Id, DocumentChunk> result : results) {
+            double score = RelevanceScore.fromCosineSimilarity(1.0f - result.getDistance());
+            if (checkMin && score < minScore) {
+                continue;
+            }
+
+            matches.add(embeddingMatch(score, result.getKey(), result.getValue()));
+        }
+
+        matches.sort(COMPARATOR);
+        return new EmbeddingSearchResult<>(matches.size() > request.maxResults()
+                                           ? matches.subList(0, request.maxResults())
```

Comment:
Do I understand correctly that `results` will contain **_all_** the chunks that were ever added? And filtering by `minScore` and `maxResults` happens here? Isn't it very inefficient?

Reply:
No, we pass `request.maxResults()` as an argument to `SimilaritySearch` aggregator, which is what performs the actual search across the cluster in parallel, and returns only the specified number of best matches.

I believe the above was to handle the case that I removed during last review cycle, which was adding previous and next chunk as well for "highly relevant" results, which would've increased the total number of results and pushed some out, so we had to trim the list of matches at the end.

I don't think that is necessary any longer, and I don't think we need to sort the matches either, as they should already be sorted by the `SimilaritySearch` aggregator. I'll simplify/remove the code above, good catch.